File: Algo/Leetcode/086PartitionList.cpp
//1. Use two List to store smaller ones and the bigger ones. Merge Them together at last. O(n)
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        if(!head) return NULL;
        ListNode* l = new ListNode(0);
        ListNode* r = new ListNode(0);
        ListNode* l_move = l;
        ListNode* r_move = r;
        while(head){
            if(head->val < x){
                l_move -> next = head;
                l_move = l_move -> next;
            } else{
                r_move->next = head;
                r_move = r_move -> next;
            }
            head = head->next;
        }
        l_move -> next = r->next;
        r_move -> next = NULL;
        return l -> next;
    }
};

File: Algo/Leetcode/086PartitionList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        if head is None:
            return head

        l = ListNode()
        r = ListNode()
        l_m = l
        r_m = r
        while head:
            if head.val < x:
                l_m.next = head
                l_m = l_m.next
            else:
                r_m.next = head
                r_m = r_m.next
            head = head.next

        # relink and stop
        l_m.next = r.next
        r_m.next = None

        return l.next